src/main.cpp

```diff
@@ -6802,108 +6802,121 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
-        pfrom->nServices = nServices;
-        if (pfrom->nVersion < MIN_PEER_PROTO_VERSION)
-        {
-            // disconnect from peers older than this proto version
-            LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
-            pfrom->PushMessage("reject", strCommand, REJECT_OBSOLETE,
-                               strprintf("Version must be %d or greater", MIN_PEER_PROTO_VERSION));
+        // Initialize all temporaries defaults
+        int32_t nVersion{0};
+        uint64_t nServices{0};
+        int64_t nTime{0};
+        CAddress addrMe{};
+        CAddress addrFrom{};
+        uint64_t nNonce{1U};
+        std::string strSubVer{};
+        std::string cleanSubVer{};
+        int32_t nStartingHeight{0};
+        bool fRelay{true};
+
+        // Consume all available payload
+        // This block might throw an IO exception. If that's the case then
+        // nothing in pfrom's state will be changed.
+        vRecv >> nVersion >> nServices >> nTime >> addrMe;
+        if (nVersion >= 106) {
+            if (vRecv.in_avail() < 39) {
+                // Be strict about size of addr_from + nonce + user_agent + start_height
+                pfrom->PushMessage("reject", strCommand, REJECT_MALFORMED, string("Error in message size"));
+                pfrom->fDisconnect = true;
+                return false;
+            }
+            vRecv >> addrFrom >> nNonce >> LIMITED_STRING(strSubVer, MAX_SUBVERSION_LENGTH) >> nStartingHeight;
+            cleanSubVer = SanitizeString(strSubVer, SAFE_CHARS_SUBVERSION);
+            if (nVersion >= 70001) {
+                if (vRecv.in_avail() < 1) {
+                    // Be strict about relay size
+                    pfrom->PushMessage("reject", strCommand, REJECT_MALFORMED, string("Error in message size"));
+                    pfrom->fDisconnect = true;
+                    return false;
+                }
+                vRecv >> fRelay;
+            }
+        }
+
+        // Should we still have unconsumed leftover in payload's datastream the message
+        // have been deliberately bloated on sender's side
+        if (vRecv.in_avail() > 0) {
+            pfrom->PushMessage("reject", strCommand, REJECT_MALFORMED, string("Error in message size"));
             pfrom->fDisconnect = true;
             return false;
         }
 
-        if (chainparams.NetworkIDString() == "test" &&
-            pfrom->nVersion < MIN_TESTNET_PEER_PROTO_VERSION)
-        {
-            // disconnect from testnet peers older than this proto version
-            LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
+        // Validate Payload
+
+        // Minimum allowed protocol version
+        const int32_t nVersion_min{chainparams.NetworkIDString() == "test" ? MIN_TESTNET_PEER_PROTO_VERSION : MIN_PEER_PROTO_VERSION};
+        if (nVersion < nVersion_min) {
```

Comment:
This is an unnecessary refactoring.

Reply:
I thought for readibility would have been good to remove two roughly identical code blocks which differ only for the value of a single variable being tested against.